src/cryptonote_core/stake_transaction_processor.cpp

```diff
@@ -0,0 +1,127 @@
+#include "stake_transaction_processor.h"
+
+using namespace cryptonote;
+
+namespace
+{
+
+const char* STAKE_TRANSACTION_STORAGE_FILE_NAME = "stake_transactions.bin";
+
+}
+
+StakeTransactionProcessor::StakeTransactionProcessor(Blockchain& blockchain)
+  : m_blockchain(blockchain)
+  , m_storage(STAKE_TRANSACTION_STORAGE_FILE_NAME)
+  , m_stake_transactions_need_update(true)
+{
+}
+
+void StakeTransactionProcessor::process_block(uint64_t block_index, const block& block, bool update_storage)
+{
+  if (block_index <= m_storage.get_last_processed_block_index())
+    return;
+
+  BlockchainDB& db = m_blockchain.get_db();
+
+  stake_transaction stake_tx;
+
+  for (const crypto::hash& tx_hash : block.tx_hashes)
+  {
+    const transaction& tx = db.get_tx(tx_hash);
+
+    if (!get_graft_stake_tx_extra_from_extra(tx, stake_tx.supernode_public_id, stake_tx.supernode_public_address, stake_tx.tx_secret_key))
+      continue;
+
+    stake_tx.block_height = block_index;
+    stake_tx.hash = tx.hash;
+    stake_tx.unlock_time = tx.unlock_time;
+
+    m_storage.add_tx(stake_tx);
+
+    m_stake_transactions_need_update = true;
+
+    std::stringstream ss;
+    ss << "New stake transaction found at block #" << block_index << ", tx_hash=" << stake_tx.hash << ", supernode_public_id '" << stake_tx.supernode_public_id << "'";
+    MCLOG(el::Level::Info, "global", ss.str());
```

Comment:
Why use a separate `stringstream` here instead of just left-shifting the values directly via the MCLOG macro?

Reply:
agree with @jagerman

Reply:
Fixed